src/main/java/com/gregtechceu/gtceu/common/machine/multiblock/electric/DistillationTowerMachine.java

```diff
@@ -0,0 +1,358 @@
+    }
+
+    @Override
+    public void onStructureFormed() {
+        getDefinition().setPartSorter(Comparator.comparingInt(p -> p.self().getPos().getY()));
+        getDefinition().setAllowExtendedFacing(false);
+        super.onStructureFormed();
+        var parts = getParts().stream()
+                .filter(part -> PartAbility.EXPORT_FLUIDS.isApplicable(part.self().getBlockState().getBlock()))
+                .toList();
+
+        if (!parts.isEmpty()) {
+            // Loop from controller y + offset -> highest output hatch
+            int y = getPos().getY() + yOffset;
+            int maxY = parts.get(parts.size() - 1).self().getPos().getY();
+            fluidOutputs = new ObjectArrayList<>(maxY - y);
+            for (int outputIndex = 0; y <= maxY; ++y) {
+                if (parts.size() <= outputIndex) {
+                    fluidOutputs.add(VoidFluidHandler.INSTANCE);
+                    continue;
+                }
+
+                var part = parts.get(outputIndex);
+                if (part.self().getPos().getY() == y) {
+                    part.getRecipeHandlers().stream()
+                            .filter(IFluidHandler.class::isInstance)
+                            .findFirst()
+                            .ifPresentOrElse(h -> {
+                                fluidOutputs.add((IFluidHandler) h);
+                                if (firstValid == null) firstValid = (IFluidHandler) h;
+                            },
+                                    () -> fluidOutputs.add(VoidFluidHandler.INSTANCE));
+                    outputIndex++;
+                } else if (part.self().getPos().getY() > y) {
+                    fluidOutputs.add(VoidFluidHandler.INSTANCE);
+                } else {
+                    GTCEu.LOGGER.error(
+                            "The Distillation Tower at {} has a fluid export hatch with an unexpected Y position",
+                            getPos());
+                    onStructureInvalid();
+                    return;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void onStructureInvalid() {
+        fluidOutputs = null;
+        firstValid = null;
+        super.onStructureInvalid();
+    }
+
+    public int limitParallel(GTRecipe recipe, int multiplier) {
+        int minMultiplier = 0;
+        int maxMultiplier = multiplier;
+
+        var maxAmount = recipe.getOutputContents(FluidRecipeCapability.CAP).stream()
+                .map(Content::getContent)
+                .map(FluidRecipeCapability.CAP::of)
+                .filter(i -> !i.isEmpty())
+                .map(i -> i.getStacks()[0])
+                .mapToInt(FluidStack::getAmount)
+                .max()
+                .orElse(0);
+
+        if (maxAmount == 0) return multiplier;
+
+        while (minMultiplier != maxMultiplier) {
+            if (multiplier > Integer.MAX_VALUE / maxAmount) multiplier = Integer.MAX_VALUE / maxAmount;
```

Comment:
Would recommend adding a linebreak for easier readability

```suggestion
            if (multiplier > Integer.MAX_VALUE / maxAmount)
                multiplier = Integer.MAX_VALUE / maxAmount;
            
```